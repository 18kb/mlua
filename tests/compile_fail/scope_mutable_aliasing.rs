use mlua::{Lua, UserData, Result};

struct MyUserData<'a>(&'a mut i32);
impl<'a> UserData for MyUserData<'a> {}

fn main() {
    let mut i = 1;

    let lua = Lua::new();
    lua.scope(|scope| -> Result<()> {
        let _a = scope.create_nonstatic_userdata(MyUserData(&mut i))?;
        let _b = scope.create_nonstatic_userdata(MyUserData(&mut i))?;
        Ok(())
    });
}
