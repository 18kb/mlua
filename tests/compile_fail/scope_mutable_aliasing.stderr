error[E0499]: cannot borrow `i` as mutable more than once at a time
  --> $DIR/scope_mutable_aliasing.rs:12:61
   |
11 |         let _a = scope.create_nonstatic_userdata(MyUserData(&mut i))?;
   |                                                             ------ first mutable borrow occurs here
12 |         let _b = scope.create_nonstatic_userdata(MyUserData(&mut i))?;
   |                        -------------------------            ^^^^^^ second mutable borrow occurs here
   |                        |
   |                        first borrow later used by call
